bot.py

```diff
@@ -133,6 +166,144 @@ async def info(self, ctx: commands.Context[Any]) -> None:
         await ctx.reply(embed=constants.INFO_REPLY)
 
 
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+def only_from_guilds(*servers: list[int]) -> Callable[[Callable[P, T]], Callable[P, T]]:
+    """Add a check to make sure the command can only be run from one of a list of servers."""
+
+    # The lazy return type is due to discord.py not being more specific in their return type.
+    # `discord.Interaction` is currently a generic type, but is not documented that way.
+
+    def check_guild(itr: discord.Interaction) -> bool:  # type: ignore[type-arg]
+        return itr.guild_id in (servers or [])
+
+    return app_commands.check(check_guild)
+
+
+class ModCommands(commands.Cog):
+    __slots__ = ("bot",)
+
+    def __init__(self, sbot: MyBot) -> None:
+        self.bot = sbot
+
+    @app_commands.command()
+    @app_commands.default_permissions(manage_messages=True)
+    @only_from_guilds(*settings.discord.management_servers)
+    async def user_code(
+        self,
+        interaction: discord.Interaction,  # type: ignore[type-arg]
+        user: discord.User,
+        day: Annotated[AdventDay, app_commands.Range[int, 1, 25]],
+        part: Annotated[AdventPart, Literal[1, 2]],
+    ) -> None:
+        await interaction.response.send_message(content="Loading...")
+
+        logger.info(
+            "User %s requested submissions for user %s, day %s, part %s",
+            interaction.user,
+            user,
+            day,
+            part,
+        )
+
+        with Database() as db:
+            results = db.get_user_submissions(lib.year(), day, part, user.id)
+
+        results.sort(key=(lambda r: r.average_time or Picoseconds(5e12)))
+        results = results[:10]
+        results.sort(key=(lambda r: r.id))
+
+        attachments = []
+        for res in results:
+            name = f"Submission_{res.id}.rs"
+            desc = f"Submission {res.id} from user {user}"
+            file_handle = StringIO(res.code)
+
+            # Not sure whether the type is specified incorrectly or this is an ugly hack.
+            # Either way, this is what works to get Discord to not open a file.
+            f = discord.File(file_handle, filename=name, description=desc)  # type: ignore[arg-type]
+            attachments.append(f)
+
+        await interaction.edit_original_response(
+            content=f"Ten fastest submissions for user {user} on day {day}, part {part}:",
+            attachments=attachments,
+        )
+
+    @app_commands.command()
+    @app_commands.default_permissions(manage_messages=True)
+    @only_from_guilds(*settings.discord.management_servers)
+    async def leaderboard_code(
+        self,
+        interaction: discord.Interaction,  # type: ignore[type-arg]
+        day: Annotated[AdventDay, app_commands.Range[int, 1, 25]],
+        part: Annotated[AdventPart, Literal[1, 2]],
+        place: Optional[app_commands.Range[int, 1, 10]] = None,
```

Comment:
Open to suggestions on what should be here. Leaning towards not having user passed in, though, since the other command covers that.